File: windows/Max.Core/Channels/TelegramChannel.cs
using System.Text;
using System.Text.Json.Nodes;

namespace Max.Core;

/// <summary>
/// Telegram bot channel — long-poll getUpdates, reply with sendMessage. Allowlist is
/// the user's NUMERIC id (from @userinfobot), never the bot's username.
/// </summary>
public sealed class TelegramChannel
{
    private readonly ISecretStore _secrets;
    private readonly ChannelRouter _router;
    private static readonly HttpClient Http = new() { Timeout = Timeout.InfiniteTimeSpan };
    private CancellationTokenSource? _cts;
    private long _offset;

    public TelegramChannel(ISecretStore secrets, ChannelRouter router) { _secrets = secrets; _router = router; }

    public void Start()
    {
        var config = MaxConfig.Load();
        var token = _secrets.Get("telegram-bot");
        if (!config.TelegramEnabled || string.IsNullOrEmpty(token)) return;
        if (_cts != null) return;
        _cts = new CancellationTokenSource();
        _ = PollLoopAsync(token!, _cts.Token);
    }

    public void Stop() { _cts?.Cancel(); _cts = null; }

    private async Task PollLoopAsync(string token, CancellationToken ct)
    {
        await PrimeOffsetAsync(token, ct);
        ChannelLog.Write("telegram: connected, polling");
        while (!ct.IsCancellationRequested)
        {
            var updates = await GetUpdatesAsync(token, 30, ct);
            if (updates is null) { try { await Task.Delay(3000, ct); } catch { } continue; }
            foreach (var u in updates)
            {
                if ((long?)u?["update_id"] is long id) _offset = Math.Max(_offset, id + 1);
                await HandleAsync(u!.AsObject(), token, ct);
            }
        }
    }

    private async Task PrimeOffsetAsync(string token, CancellationToken ct)
    {
        var updates = await GetUpdatesAsync(token, 0, ct);
        if (updates is null) return;
        foreach (var u in updates)
            if ((long?)u?["update_id"] is long id) _offset = Math.Max(_offset, id + 1);
    }

    private async Task<JsonArray?> GetUpdatesAsync(string token, int timeout, CancellationToken ct)
    {
        try
        {
            var url = $"https://api.telegram.org/bot{token}/getUpdates?offset={_offset}&timeout={timeout}";
            using var resp = await Http.GetAsync(url, ct);
            var body = await resp.Content.ReadAsStringAsync(ct);
            var obj = JsonNode.Parse(body)?.AsObject();
            if (obj?["ok"]?.GetValue<bool>() != true)
            {
                ChannelLog.Write($"telegram: API error {obj?["description"]} — check the bot token");
                return null;
            }
            return obj!["result"] as JsonArray ?? new JsonArray();
        }
        catch (OperationCanceledException) { return null; }
        catch (Exception ex) { ChannelLog.Write($"telegram: poll error {ex.Message}"); return null; }
    }

    private async Task HandleAsync(JsonObject update, string token, CancellationToken ct)
    {
        var message = update["message"]?.AsObject();
        var text = (string?)message?["text"];
        var from = message?["from"]?.AsObject();
        var chat = message?["chat"]?.AsObject();
        if (message is null || string.IsNullOrWhiteSpace(text) || from is null || chat is null) return;
        if (from["is_bot"]?.GetValue<bool>() == true) return;

        var userId = ((long?)from["id"])?.ToString() ?? "0";
        var chatId = ((long?)chat["id"])?.ToString() ?? "0";
        if (userId == "0" || chatId == "0") return;

        var allow = MaxConfig.Load().TelegramAllowlist;
        if (!allow.Contains(userId))
        {
            ChannelLog.Write($"telegram: ignored message from {(string?)from["username"] ?? "?"} (id {userId}) — not in allowlist");
            return;
        }

        ChannelLog.Write($"telegram: handling message from {userId}");
        await _router.HandleAsync($"telegram-{chatId}", text!.Trim(),
            reply => SendMessageAsync(chatId, reply, token), ct);
    }

    private static async Task SendMessageAsync(string chatId, string text, string token)
    {
        try
        {
            var body = new JsonObject { ["chat_id"] = chatId, ["text"] = text.Length > 4000 ? text[..4000] : text };
            using var content = new StringContent(body.ToJsonString(), Encoding.UTF8, "application/json");
            await Http.PostAsync($"https://api.telegram.org/bot{token}/sendMessage", content);
        }
        catch (Exception ex) { ChannelLog.Write($"telegram: send error {ex.Message}"); }
    }
}

File: Sources/AskMax/Core/Channels/TelegramChannel.swift
import Foundation

/// Telegram bot — the simplest channel: long-poll getUpdates (no WebSocket, no
/// public webhook), reply with sendMessage. Token comes from @BotFather.
@MainActor
final class TelegramChannel {
    static let shared = TelegramChannel()

    private var running = false
    private var token = ""
    private var offset = 0
    private var pollTask: Task<Void, Never>?

    private init() {}

    func startIfEnabled() {
        let config = MaxConfig.load()
        guard config.telegramEnabled,
              let t = SecretStore.secret(forKey: "telegram-bot"), !t.isEmpty else {
            stop(); return
        }
        guard !running else { return }
        token = t
        running = true
        pollTask = Task { [weak self] in
            await self?.primeOffset()  // skip messages sent before we came online
            await self?.pollLoop()
        }
    }

    func stop() {
        running = false
        pollTask?.cancel()
        pollTask = nil
    }

    /// Fast initial call to advance past any backlog so we don't reply to old
    /// messages on startup.
    private func primeOffset() async {
        guard let results = await fetchUpdates(timeout: 0) else { return }
        for update in results {
            if let id = update["update_id"] as? Int { offset = max(offset, id + 1) }
        }
    }

    private func pollLoop() async {
        while running && !Task.isCancelled {
            guard let results = await fetchUpdates(timeout: 30) else {
                try? await Task.sleep(nanoseconds: 3_000_000_000) // back off on error
                continue
            }
            for update in results {
                if let id = update["update_id"] as? Int { offset = max(offset, id + 1) }
                handleUpdate(update)
            }
        }
    }

    private func fetchUpdates(timeout: Int) async -> [[String: Any]]? {
        var comps = URLComponents(string: "https://api.telegram.org/bot\(token)/getUpdates")!
        comps.queryItems = [
            URLQueryItem(name: "offset", value: String(offset)),
            URLQueryItem(name: "timeout", value: String(timeout)),
        ]
        guard let url = comps.url else { return nil }
        var request = URLRequest(url: url)
        request.timeoutInterval = TimeInterval(timeout + 15)
        do {
            let (data, _) = try await URLSession.shared.data(for: request)
            guard let obj = JSON.decode(data), obj["ok"] as? Bool == true,
                  let results = obj["result"] as? [[String: Any]] else {
                if let obj = JSON.decode(data), obj["ok"] as? Bool == false {
                    ChannelLog.write("telegram: API error \(obj["description"] ?? "?") — check the bot token")
                }
                return nil
            }
            return results
        } catch {
            return nil
        }
    }

    private func handleUpdate(_ update: [String: Any]) {
        guard let message = update["message"] as? [String: Any],
              let text = (message["text"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines),
              !text.isEmpty,
              let from = message["from"] as? [String: Any],
              let chat = message["chat"] as? [String: Any] else { return }

        if from["is_bot"] as? Bool == true { return }
        let userId = String(describing: from["id"] as? Int ?? 0)
        let chatId = String(describing: chat["id"] as? Int ?? 0)
        guard userId != "0", chatId != "0" else { return }

        let allow = MaxConfig.load().telegramAllowlist
        guard allow.contains(userId) else {
            let name = from["username"] as? String ?? "?"
            ChannelLog.write("telegram: ignored message from \(name) (id \(userId)) — not in allowlist")
            return
        }

        let token = self.token
        ChannelRouter.shared.handle(key: "telegram-\(chatId)", text: text) { reply in
            Self.sendReply(chatId: chatId, text: reply, token: token)
        }
    }

    nonisolated static func sendReply(chatId: String, text: String, token: String) {
        var request = URLRequest(url: URL(string: "https://api.telegram.org/bot\(token)/sendMessage")!)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = JSON.encode(["chat_id": chatId, "text": String(text.prefix(4000))])
        URLSession.shared.dataTask(with: request).resume()
    }
}
